### PR TITLE
Fix type inference for simple case expression operand

### PR DESCRIPTION
**Pull request by wenshao**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -558,9 +558,9 @@ protected Type visitSearchedCaseExpression(SearchedCaseExpression node, Stackabl
         @Override
         protected Type visitSimpleCaseExpression(SimpleCaseExpression node, StackableAstVisitorContext<Context> context)
         {
-            for (WhenClause whenClause : node.getWhenClauses()) {
-                coerceToSingleType(context, whenClause, "CASE operand type does not match WHEN clause operand type: %s vs %s", node.getOperand(), whenClause.getOperand());
-            }
+            coerceToSingleType(context,
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java]: The existing code already checks that every operand of a `when` clause can be coerced to a common type with the operand of the `case` clause. Can you give an example of what issue this is intending to fix?

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java [hunk 2]
```diff
@@ -442,4 +442,15 @@ public void testSimpleCase()
...
+
+    @Test
+    public void testSimpleCase1()
+    {
+        assertFunction("case 8 " +
+                        "when double '76.1' then double '1.0' " +
+                        "when real '8.0' then null " +
+                        "end",
+                DOUBLE,
+                null);
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java]: Note: the test passes before applying the code changes here.

```
presto:default> SELECT case 8 when double '76.1' then double '1.0' when real '8.0' then null end;
 _col0
-------
  NULL
(1 row)
```

```
presto:default> SELECT typeof(case 8 when double '76.1' then double '1.0' when real '8.0' then null end);
 _col0
--------
 double
(1 row)
```

Comment (wenshao) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java]:
```sql
SELECT CASE 8 when 9 then 1 when cast(null as decimal) then 3 else 4 end 
```

Comment (findepi) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java]: Indeed.
```
presto:default> SELECT CASE 8 when 9 then 1 when cast(null as decimal(20,1)) then 3 else 4 end ;

Query 20191022_130929_00006_x3p8n failed: Cannot cast java.lang.Long to io.airlift.slice.Slice
java.lang.ClassCastException: Cannot cast java.lang.Long to io.airlift.slice.Slice
	at java.base/java.lang.Class.cast(Class.java:3605)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:735)
	at io.prestosql.sql.InterpretedFunctionInvoker.invoke(InterpretedFunctionInvoker.java:88)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.invokeOperator(ExpressionInterpreter.java:1298)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.isEqual(ExpressionInterpreter.java:515)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitSimpleCaseExpression(ExpressionInterpreter.java:498)
	at io.prestosql.sql.tree.SimpleCaseExpression.accept(SimpleCaseExpression.java:70)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.planner.ExpressionInterpreter.optimize(ExpressionInterpreter.java:300)
	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.rewrite(SimplifyExpressions.java:53)
	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.lambda$createRewrite$0(SimplifyExpressions.java:78)
	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$ProjectExpressionRewrite.lambda$apply$0(ExpressionRewriteRuleSet.java:125)
	at io.prestosql.sql.planner.plan.Assignments.lambda$rewrite$1(Assignments.java:108)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.prestosql.sql.planner.plan.Assignments.rewrite(Assignments.java:109)
	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$ProjectExpressionRewrite.apply(ExpressionRewriteRuleSet.java:125)
	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$ProjectExpressionRewrite.apply(ExpressionRewriteRuleSet.java:106)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.transform(IterativeOptimizer.java:165)
...
```

File: presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java (hunk 1 above)
Comment (wenshao) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java]: operand is 8, first whenClause coerce operand 8 to double type, second whenCaluse coerce operand 8 to real type, overwrite previous coerce type.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java (hunk 2 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java]: Thanks. Can you add that case to the test, instead?

File: presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -558,9 +558,9 @@ protected Type visitSearchedCaseExpression(SearchedCaseExpression node, Stackabl
...
+                    "All CASE results must be the same type: %s",
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java]: In this case, we want the error message to indicate *which* `when` operand cannot be compared to the `case` operand. So the fix is a little more involved. We need to attempt to find the common supertype between each `when` operand and the `case` operand and report an error if necessary. As we do this, keep track of the common supertype between all the types seen so far. Then, at the end, add a coercion for the case operand and all when operands to the common supertype. To illustrate:

```
for each when_operand in when_clauses
    t = supertype(when_operand, case_operand)
    if t not found => error
    common_type_so_far = supertype(t, common_type_so_far)   // or just t in the first loop

add_coercion(case_operand, common_type_so_far)
for each when_operand in when_clauses
    add_coercion(when_operand, common_type_so_far)
```

Comment (wenshao) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java]: it has been modified according to your suggestion

**Comment by martint** [comment]
Thanks! Please make sure to submit a signed CLA so we can merge this change: https://github.com/prestosql/cla

**Comment by findepi** [comment]
@wenshao please see the Travis output. `TestExpressionInterpreter` is failing.

**Comment by martint** [comment]
@cla-bot check

**Comment by wenshao** [comment]
> @wenshao please see the Travis output. `TestExpressionInterpreter` is failing.



> @wenshao please see the Travis output. `TestExpressionInterpreter` is failing.

fixed

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java [hunk 4]
```diff
@@ -442,4 +442,22 @@ public void testSimpleCase()
                 DOUBLE,
                 33.0);
     }
+
+    @Test
+    public void testSimpleCase1()
```
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java]: Name this `testSimpleCaseWithCoercions`

File: presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -575,6 +573,50 @@ protected Type visitSimpleCaseExpression(SimpleCaseExpression node, StackableAst
             return type;
         }
 
+        private Type coerceCaseOperandToToSingleType(SimpleCaseExpression node, StackableAstVisitorContext<Context> context)
+        {
+            Type operandType = process(node.getOperand(), context);
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java]: This implementation can be simplified a bit. Here's what I came up with after playing with the code a bit:

```java
Type operandType = process(node.getOperand(), context);
List<Type> whenOperandTypes = new ArrayList<>();

Type globalCommonType = null;
for (WhenClause whenClause : node.getWhenClauses()) {
    Type type = process(whenClause.getOperand(), context);
    whenOperandTypes.add(type);

    Optional<Type> commonType = typeCoercion.getCommonSuperType(operandType, type);

    if (!commonType.isPresent()) {
        throw semanticException(TYPE_MISMATCH, whenClause.getOperand(), "CASE operand type does not match WHEN clause operand type: %s vs %s", operandType, type);
    }

    if (globalCommonType == null) {
        globalCommonType = commonType.get();
    }

    globalCommonType = typeCoercion.getCommonSuperType(globalCommonType, commonType.get()).get();
}

coerceType(node.getOperand(), operandType, globalCommonType, "CASE operand");
for (int i = 0; i < node.getWhenClauses().size(); i++) {
    coerceType(node.getWhenClauses().get(i).getOperand(), whenOperandTypes.get(i), globalCommonType, "WHEN clause operand");
}

return operandType;
```

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java (hunk 4 above)
Comment (wenshao) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestConditions.java]: done

**Comment by martint** [comment]
Merged! Thanks for your contribution!

**Pull request closed without merging**